js/web/lib/onnxjs/graph.ts

```diff
@@ -585,33 +585,48 @@ class GraphImpl implements Graph, Graph.Transformer {
   finalizeGraph() {
     let offset = 0;
     // delete all nodes that are not being executed
+    // The graph is represented using these two arrays
+    // this._nodes - Array holding the kernels to execute - each entry is a kernel pointing to this._allData
+    // this._allData - hold 2 fields - to [] & from [] - these arrays hold the graph map for inputs and outputs per node
+    // reducedMap - remapping the graph after reading the flag 'executeNode' 
+    const reducedGraph = new Array (this._nodes.length,0);
+    let nodePossition = 0;
+
     for (let i = 0; i < this._nodes.length; i++) {
-      if (!this._nodes[i].executeNode) {
-        // delete this node and shift all subsequent nodes up
-        offset++;
+      // giving new indexes to the nodes based on execution flag
+      reducedGraph[i] = nodePossition;
+      if (this._nodes[i].executeNode) { 
+        
+        if(nodePossition!=i){
+          this._nodes[nodePossition] = this._nodes[i];
+        }
+
+        nodePossition++;
+      }else{
         // delete all output values
         this._nodes[i].outputs.forEach(ind => {
           this._allData[ind]._from = -2;
         });
-        this._nodes.splice(i, 1);
-        i--;
-        continue;
       }
-      if (offset > 0) {
-        // update the value table
-        this._nodes[i].inputs.forEach(value => {
-          const ind = this._allData[value]._to.indexOf(i + offset);
-          if (ind !== -1) {
-            this._allData[value]._to[ind] = i;
-          }
-        });
-        this._nodes[i].outputs.forEach(value => {
-          if (this._allData[value]._from && this._allData[value]._from! === i + offset) {
-            this._allData[value]._from! = i;
-          }
-        });
+    }
+
+    // removing the unused nodes
+    this._nodes.splice(nodePossition,this._nodes.length - nodePossition);
+
+    // Updating this._allData according to the new this._nodes
+    for (let i = 0; i < this._allData.length; i++) {
+
```

Comment:
add a local variable because you are going to access it mutiple times

`const data = this._allData[i];`